packages/clay-core/src/side-panel/SidePanel.tsx

```diff
@@ -0,0 +1,189 @@
+	 * only important content.
+	 */
+	as?: 'aside' | 'nav' | 'section';
+
+	/**
+	 * Sets the CSS className for the component.
+	 */
+	className?: string;
+
+	/**
+	 * Children content to render a content.
+	 */
+	children: React.ReactNode;
+
+	/**
+	 * Property to set the default value (uncontrolled).
+	 */
+	defaultOpen?: boolean;
+
+	/**
+	 * Direction the panel will render.
+	 */
+	direction?: 'left' | 'right';
+
+	/**
+	 * Flag to determine which style the SidePanel will display.
+	 */
+	displayType?: 'light' | 'dark';
+
+	/**
+	 * The id of the component.
+	 */
+	id?: string;
+
+	/**
+	 * Element reference to open the SidePanel.
+	 * NOTE: SidePanel automatically identifies the trigger but it may not
+	 * work in some cases when the element disappears from the DOM.
+	 */
+	triggerRef?: React.RefObject<HTMLElement>;
+} & (UncontrolledState | ControlledState);
+
+export function SidePanel({
+	as: As = 'aside',
+	children,
+	className,
+	defaultOpen,
+	direction = 'right',
+	displayType = 'light',
+	onOpenChange,
+	open: externalOpen,
+	triggerRef,
+	...otherProps
+}: Props) {
+	const ref = useRef<HTMLElement | null>(null);
+
+	const [open, setOpen] = useControlledState({
+		defaultName: 'defaultOpen',
+		defaultValue: defaultOpen,
+		handleName: 'onOpenChange',
+		name: 'open',
+		onChange: onOpenChange,
+		value: externalOpen,
+	});
+
+	const [transition, setTransition] = useState(false);
+
+	useEffect(() => {
+		if (open) {
+			// Saves the last element that has focus before moving to the sidepanel to
+			// regain focus when closing the sidepanel.
+			const element =
+				(document.activeElement as HTMLElement | null) ||
+				triggerRef?.current ||
+				null;
+
+			return () => {
+				element?.focus();
+			};
+		}
+	}, [open]);
+
+	useEffect(() => {
+		if (open) {
+			const onKeyDown = (event: KeyboardEvent) => {
+				if (event.key === Keys.Esc) {
+					event.stopImmediatePropagation();
+					event.preventDefault();
+
+					setOpen(false);
+				}
+			};
+
+			// Move focus to sidepanel when opening
+			ref.current?.focus();
+
+			document.addEventListener('keydown', onKeyDown, true);
+
+			return () => {
+				document.removeEventListener('keydown', onKeyDown, true);
+			};
+		}
+	}, [open]);
+
+	useLayoutEffect(() => {
+		setTransition(true);
+
+		return () => {
+			setTransition(false);
+		};
+	}, [open]);
+
+	return (
+		<div
+			className={classnames('c-slideout c-slideout-fixed', {
+				'c-slideout-end': direction === 'right',
+				'c-slideout-shown': !transition && open,
+				'c-slideout-start': direction === 'left',
+			})}
+		>
+			<As
+				className={classnames('sidebar c-slideout-active', className, {
+					'c-slideout-show': open,
+					'c-slideout-transition': transition,
+					'c-slideout-transition-in': transition && open,
+					'c-slideout-transition-out': transition && !open,
+					'sidebar-dark': displayType === 'dark',
+					'sidebar-light': displayType === 'light',
+				})}
+				onTransitionEnd={() => setTransition(false)}
```

Comment:
I wanted to move this animation to [motion](https://motion.dev/) which handles animations better and also reduced motion for example but unfortunately it is only compatible with React v19. I need to check if old versions are compatible with React v16.

Reply:
@matuzalemsteles the slide in isn't animating because there needs to be a slight delay for the browser to apply `c-slideout-transition c-slideout-transition-in` styles. I made some little tweaks and it's working for me, but it's hacky.

https://github.com/pat270/clay/commit/55216977e43447042c728323a5deb794791cafaf

Reply:
Thanks @pat270! I'm going to try using `motion` for this, I think it will simplify and handle edge cases better.